File: idor_lens.php
#!/usr/bin/env php
<?php
/** IDOR Lens: PHP 8.1+ CLI with ext-curl; same JSON config as Python. */
declare(strict_types=1);
const BODY_LIMIT = 1048576;
function fail(string $message): never { fwrite(STDERR, $message . PHP_EOL); exit(2); }
function pointer(mixed $data, string $path): mixed {
    foreach (explode('/', substr($path, 1)) as $part) {
        $part = str_replace(['~1', '~0'], ['/', '~'], $part);
        if (is_object($data) && property_exists($data, $part)) $data = $data->$part;
        elseif (is_array($data) && ctype_digit($part) && array_key_exists((int)$part, $data)) $data = $data[(int)$part];
        else return null;
    }
    return $data;
}
function matches(array $r, string $pointer, mixed $expected): bool {
    $value = pointer($r['json'] ?? null, $pointer);
    return !($r['truncated'] ?? false) && ($r['status'] ?? 0) >= 200 && ($r['status'] ?? 0) < 300
        && (is_string($value) || is_int($value)) && (string)$value === (string)$expected;
}
function evidence(array $r): array { unset($r['json']); return $r; }
function fetch(string $url, array $headers, float $rate): array {
    static $last = 0.0;
    $now = hrtime(true) / 1e9;
    usleep((int)(max(0, 1 / $rate - ($now - $last)) * 1e6));
    $last = hrtime(true) / 1e9;
    $body = ''; $truncated = false;
    $ch = curl_init($url);
    curl_setopt_array($ch, [CURLOPT_HTTPHEADER => array_merge(['Accept: application/json'], $headers),
        CURLOPT_FOLLOWLOCATION => false, CURLOPT_PROXY => '', CURLOPT_CONNECTTIMEOUT => 10,
        CURLOPT_TIMEOUT => 10, CURLOPT_SSL_VERIFYPEER => true, CURLOPT_SSL_VERIFYHOST => 2,
        CURLOPT_PROTOCOLS => CURLPROTO_HTTP | CURLPROTO_HTTPS,
        CURLOPT_WRITEFUNCTION => function ($ch, string $chunk) use (&$body, &$truncated): int {
            if (strlen($body) + strlen($chunk) > BODY_LIMIT) { $truncated = true; return 0; }
            $body .= $chunk; return strlen($chunk);
        }]);
    $ok = curl_exec($ch); $status = curl_getinfo($ch, CURLINFO_RESPONSE_CODE); curl_close($ch);
    if ($ok === false && !$truncated) return ['status' => null, 'error' => 'request_failed', 'json' => null];
    $r = ['status' => $status ?: null, 'bytes' => strlen($body), 'truncated' => $truncated, 'json' => null];
    if (!$truncated) { $r['sha256'] = hash('sha256', $body); $r['json'] = json_decode($body); }
    return $r;
}
if (PHP_SAPI !== 'cli') fail('Run this tool from the command line.');
if (!extension_loaded('curl')) fail('The PHP curl extension is required.');
if ($argc !== 2 && !($argc === 4 && $argv[2] === '--output')) fail('Usage: php idor_lens.php CONFIG [--output REPORT]');
try {
    $cfg = json_decode(@file_get_contents($argv[1]), true, 512, JSON_THROW_ON_ERROR);
    $base = $cfg['base_url'] ?? '';
    if (!preg_match('~^https?://(?:\[[0-9a-fA-F:]+\]|[A-Za-z0-9.-]+)(?::[0-9]{1,5})?/?$~D', $base)) throw new Exception();
    $u = parse_url($base);
    if ($u === false || ($u['scheme'] === 'http' && !in_array($u['host'], ['localhost','127.0.0.1','[::1]'], true))) throw new Exception();
    $rate = $cfg['requests_per_second'] ?? 1;
    if (!is_numeric($rate) || !is_finite((float)$rate) || $rate <= 0 || $rate > 10) throw new Exception();
    $accounts = $cfg['accounts'] ?? [];
    if (count($accounts) !== 2) throw new Exception();
    $names = []; $headers = []; $identities = [];
    foreach ($accounts as $a) {
        if (!is_string($a['name']) || $a['name'] === '') throw new Exception();
        $names[] = $a['name']; $h = [];
        foreach ($a['headers_from_env'] as $key => $env) {
            $key = strtolower($key); $value = getenv($env);
            if (!in_array($key, ['authorization','cookie','x-api-key'], true) || !$value || strpbrk($value, "\r\n") !== false) throw new Exception();
            $h[$key] = $value;
        }
        if (!$h) throw new Exception();
        ksort($h); $identities[] = $h;
        $headers[] = array_map(fn($k, $v) => "$k: $v", array_keys($h), array_values($h));
    }
    if ($names[0] === $names[1] || $identities[0] === $identities[1]) throw new Exception();
    if (empty($cfg['cases'])) throw new Exception();
    foreach ($cfg['cases'] as $case) {
        $path = $case['path']; $ids = $case['objects'];
        if (!is_string($path) || !str_starts_with($path, '/') || str_starts_with($path, '//') || preg_match('/[\x00-\x20\x7f#\\\\]/', $path)
            || substr_count($path, '{id}') !== 1 || !str_starts_with($case['id_pointer'], '/') || count($ids) !== 2) throw new Exception();
        foreach ($names as $name) if (!isset($ids[$name]) || (!is_string($ids[$name]) && !is_int($ids[$name])) || (string)$ids[$name] === '') throw new Exception();
        if ((string)$ids[$names[0]] === (string)$ids[$names[1]]) throw new Exception();
    }
} catch (Throwable $e) { fail('Invalid configuration or missing credential environment variable.'); }
$rows = []; $stopped = false;
foreach ($cfg['cases'] as $index => $case) {
    $request = function(int $actor, int $owner) use ($base, $case, $names, $headers, $rate): array {
        return fetch(rtrim($base, '/') . str_replace('{id}', rawurlencode((string)$case['objects'][$names[$owner]]), $case['path']), $headers[$actor], (float)$rate);
    };
    $baseline = [];
    for ($i = 0; $i < 2; $i++) {
        $baseline[$i] = $request($i, $i);
        if ($baseline[$i]['status'] === 429) { $stopped = true; break; }
    }
    if ($stopped) break;
    $valid = matches($baseline[0], $case['id_pointer'], $case['objects'][$names[0]]) && matches($baseline[1], $case['id_pointer'], $case['objects'][$names[1]]);
    foreach ([[0,1], [1,0]] as [$actor,$owner]) {
        $r = $request($actor, $owner);
        $result = !$valid ? 'inconclusive_baseline' : (matches($r, $case['id_pointer'], $case['objects'][$names[$owner]]) ? 'potential_idor' : (in_array($r['status'], [401,403,404], true) ? 'access_denied_observed' : 'inconclusive'));
        $rows[] = ['case' => $index + 1, 'actor' => $names[$actor], 'owner' => $names[$owner], 'result' => $result,
            'owner_baseline' => evidence($baseline[$owner]), 'actor_baseline' => evidence($baseline[$actor]), 'cross_account' => evidence($r)];
        printf("Case %d: %s -> %s: %s\n", $index+1, $names[$actor], $names[$owner], $result);
        if ($r['status'] === 429) { $stopped = true; break; }
    }
    if ($stopped) break;
}
$report = ['tool' => 'IDOR Lens', 'version' => '1.0.0', 'stopped_on_rate_limit' => $stopped, 'complete' => !$stopped,
    'results' => $rows, 'note' => 'Candidates require manual ownership, identity, and access-policy verification.'];
$output = $argv[3] ?? 'reports/report.json';
if (!is_dir(dirname($output)) && !@mkdir(dirname($output), 0700, true)) fail('Cannot create report directory.');
if (@file_put_contents($output, json_encode($report, JSON_PRETTY_PRINT | JSON_THROW_ON_ERROR)."\n") === false) fail('Cannot write report.');
echo "Report: $output\n";
exit($stopped ? 2 : (in_array('potential_idor', array_column($rows, 'result'), true) ? 1 : 0));

File: idor_lens.py
#!/usr/bin/env python3
"""Compare object access between two controlled accounts. Python 3.10+."""
import argparse
import hashlib
import json
import math
import os
from pathlib import Path
import time
from urllib.error import HTTPError, URLError
from urllib.parse import quote, urlsplit
from urllib.request import Request, build_opener, HTTPRedirectHandler, ProxyHandler

LIMIT = 1_048_576


class NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def lookup(document, pointer):
    """Resolve an RFC 6901 JSON pointer, including array indices."""
    if pointer == "":
        return document
    if not pointer.startswith("/"):
        raise ValueError("JSON pointers must start with /")
    for part in pointer[1:].split("/"):
        part = part.replace("~1", "/").replace("~0", "~")
        if isinstance(document, list):
            if not part.isdigit():
                raise KeyError(part)
            document = document[int(part)]
        else:
            document = document[part]
    return document


def validate(config):
    base = urlsplit(config["base_url"])
    if (base.scheme not in {"https", "http"} or not base.hostname
            or base.username or base.password or base.query or base.fragment
            or base.path not in {"", "/"}):
        raise ValueError("base_url must be an HTTP(S) origin without credentials")
    if base.scheme == "http" and base.hostname not in {"localhost", "127.0.0.1", "::1"}:
        raise ValueError("Use HTTPS for remote targets")
    rate = config.get("requests_per_second", 1)
    if not isinstance(rate, (int, float)) or not math.isfinite(rate) or not 0 < rate <= 10:
        raise ValueError("requests_per_second must be greater than 0 and at most 10")
    accounts = config["accounts"]
    if len(accounts) != 2:
        raise ValueError("Exactly two controlled accounts are required")
    names = [a["name"] for a in accounts]
    if len(set(names)) != 2:
        raise ValueError("Account names must differ")
    credentials = []
    for account in accounts:
        headers = {}
        for header, env in account["headers_from_env"].items():
            if header.lower() not in {"authorization", "cookie", "x-api-key"}:
                raise ValueError("Supported credential headers: Authorization, Cookie, X-API-Key")
            value = os.environ.get(env)
            if not value or "\n" in value or "\r" in value:
                raise ValueError("Missing or invalid credential environment variable")
            headers[header.lower()] = value
        if not headers:
            raise ValueError("Each account needs credential headers")
        credentials.append(headers)
    if credentials[0] == credentials[1]:
        raise ValueError("Account credentials must differ")
    cases = config["cases"]
    if not cases:
        raise ValueError("At least one case is required")
    for case in cases:
        path = case["path"]
        if not path.startswith("/") or path.startswith("//") or "#" in path or "\\" in path:
            raise ValueError("Case paths must be relative to base_url")
        if path.count("{id}") != 1 or any(ord(c) < 33 for c in path):
            raise ValueError("Each path needs exactly one {id} and no whitespace")
        if set(case["objects"]) != set(names):
            raise ValueError("Each case needs one known object per account")
        ids = list(case["objects"].values())
        if any(type(i) not in {str, int} or not str(i) for i in ids) or str(ids[0]) == str(ids[1]):
            raise ValueError("Object IDs must be distinct nonempty strings or integers")
        if not case["id_pointer"].startswith("/"):
            raise ValueError("id_pointer must identify an object ID field")
    return credentials


class Client:
    def __init__(self, rate):
        self.interval = 1 / rate
        self.last = 0.0
        # No ambient proxy, cookie jar, redirect following, or credential forwarding.
        self.opener = build_opener(ProxyHandler({}), NoRedirect())

    def get(self, url, headers):
        time.sleep(max(0, self.interval - (time.monotonic() - self.last)))
        self.last = time.monotonic()
        try:
            request = Request(url, headers={"Accept": "application/json", **headers})
            try:
                response = self.opener.open(request, timeout=10)
            except HTTPError as exc:
                response = exc
            with response:
                body = response.read(LIMIT + 1)
                result = {"status": response.code, "bytes": len(body),
                          "truncated": len(body) > LIMIT, "json": None}
                if len(body) <= LIMIT:
                    result["sha256"] = hashlib.sha256(body).hexdigest()
                    try:
                        result["json"] = json.loads(body)
                    except (ValueError, UnicodeError):
                        pass
                return result
        except (URLError, TimeoutError, OSError, ValueError):
            return {"status": None, "error": "request_failed", "json": None}


def matches(response, pointer, expected):
    if response.get("truncated") or not 200 <= (response["status"] or 0) < 300:
        return False
    try:
        value = lookup(response["json"], pointer)
        return type(value) in {str, int} and str(value) == str(expected)
    except (KeyError, IndexError, TypeError, ValueError):
        return False


def evidence(response):
    return {key: value for key, value in response.items() if key != "json"}


def classify(baseline_ok, response, pointer, expected):
    if not baseline_ok:
        return "inconclusive_baseline"
    if matches(response, pointer, expected):
        return "potential_idor"
    if response["status"] in {401, 403, 404}:
        return "access_denied_observed"
    return "inconclusive"


def scan(config, client=None):
    headers = validate(config)
    client = client or Client(config.get("requests_per_second", 1))
    names = [a["name"] for a in config["accounts"]]
    rows = []
    stopped = False
    for number, case in enumerate(config["cases"], 1):
        baseline = {}
        def fetch(actor, owner):
            oid = case["objects"][names[owner]]
            url = config["base_url"].rstrip("/") + case["path"].replace("{id}", quote(str(oid), safe=""))
            return client.get(url, headers[actor])
        for owner in range(2):
            baseline[owner] = fetch(owner, owner)
            if baseline[owner]["status"] == 429:
                stopped = True
                break
        if stopped:
            break
        baseline_ok = all(matches(baseline[i], case["id_pointer"], case["objects"][names[i]]) for i in range(2))
        for actor, owner in [(0, 1), (1, 0)]:
            response = fetch(actor, owner)
            rows.append({"case": number, "actor": names[actor], "owner": names[owner],
                         "result": classify(baseline_ok, response, case["id_pointer"], case["objects"][names[owner]]),
                         "owner_baseline": evidence(baseline[owner]),
                         "actor_baseline": evidence(baseline[actor]),
                         "cross_account": evidence(response)})
            if response["status"] == 429:
                stopped = True
                break
        if stopped:
            break
    return {"tool": "IDOR Lens", "version": "1.0.0", "stopped_on_rate_limit": stopped,
            "complete": not stopped, "results": rows,
            "note": "Candidates require manual ownership, identity, and access-policy verification."}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("config", type=Path)
    parser.add_argument("--output", type=Path, default=Path("reports/report.json"))
    args = parser.parse_args()
    try:
        report = scan(json.loads(args.config.read_text()))
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_text(json.dumps(report, indent=2) + "\n")
    except (OSError, ValueError, KeyError, TypeError) as exc:
        parser.exit(2, f"Configuration or output error ({type(exc).__name__}); check config and environment.\n")
    for row in report["results"]:
        print(f"Case {row['case']}: {row['actor']} -> {row['owner']}: {row['result']}")
    print(f"Report: {args.output}")
    if report["stopped_on_rate_limit"]:
        print("Stopped: server returned HTTP 429.")
        return 2
    return 1 if any(r["result"] == "potential_idor" for r in report["results"]) else 0


if __name__ == "__main__":
    raise SystemExit(main())
